### PR TITLE
Set CLOEXEC flag when opening files and sockets

### PR DESCRIPTION
**Pull request by ensc**
Avoids leaking of file descriptors in multi threaded programs by:

- setting `O_CLOEXEC` flag in `rrd_open()`
- using the "e" flag for `fopen()` and providing a fallback implementation for environments without this flag

Solves #1202

### CODE REVIEW

**Comment by ensc** [comment]
I do not know why the linux builds in the CI are failing; they succeed locally.  Perhaps a problem with the github buildsystem.

**Comment by oetiker** [comment]
looking into the test issue ... guess this might be unrelated to your changes

**Review by oetiker — APPROVED** [review]

File: src/compat-cloexec.c [hunk 1]
```diff
@@ -0,0 +1,101 @@
+#include "compat-cloexec.h"
+
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef _POSIX_C_SOURCE
```
Comment (c72578) [inline src/compat-cloexec.c:7]: @ensc Could you please check:
```c
#ifdef _POSIX_C_SOURCE
#  include <fcntl.h>
```
In msys2 mingw-w64 builds we see these warnings e.g. [1]:
```
compat-cloexec.c:11: warning: "O_CREAT" redefined
   11 | #  define O_CREAT       0
      | 
In file included from ./rrd_config_bottom.h:64,
                 from ./rrd_config.h:634,
                 from compat-cloexec.h:4,
                 from compat-cloexec.c:1:
D:/a/_temp/msys64/mingw64/include/fcntl.h:40: note: this is the location of the previous definition
   40 | #define O_CREAT _O_CREAT
      | 
compat-cloexec.c:12: warning: "O_WRONLY" redefined
   12 | #  define O_WRONLY      0
      | 
D:/a/_temp/msys64/mingw64/include/fcntl.h:37: note: this is the location of the previous definition
   37 | #define O_WRONLY _O_WRONLY
      | 
compat-cloexec.c:13: warning: "O_RDONLY" redefined
   13 | #  define O_RDONLY      0
      | 
D:/a/_temp/msys64/mingw64/include/fcntl.h:36: note: this is the location of the previous definition
   36 | #define O_RDONLY _O_RDONLY
      | 
compat-cloexec.c:14: warning: "O_APPEND" redefined
   14 | #  define O_APPEND      0
      | 
D:/a/_temp/msys64/mingw64/include/fcntl.h:39: note: this is the location of the previous definition
   39 | #define O_APPEND _O_APPEND
      | 
compat-cloexec.c:15: warning: "O_RDWR" redefined
   15 | #  define O_RDWR        0
      | 
D:/a/_temp/msys64/mingw64/include/fcntl.h:38: note: this is the location of the previous definition
   38 | #define O_RDWR _O_RDWR
      | 
```
**`configure`** says:
```
checking for fcntl.h... yes
```
So, `HAVE_FCNTL_H` is defined.

[1] https://github.com/oetiker/rrdtool-1.x/actions/runs/4740218558/jobs/8415805010#step:5:901

File: tests/test_compat-cloexec.c [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#include <compat-cloexec.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+#ifdef _POSIX_C_SOURCE
```
Comment (c72578) [inline tests/test_compat-cloexec.c:7]: @ensc Warnings under msys2 mingw-w64 as in `compat-cloexec.c`:
```sh
  CC       test_compat-cloexec.o
test_compat-cloexec.c:10: warning: "O_RDONLY" redefined
   10 | #  define O_RDONLY      0
      | 
In file included from ../src/rrd_config_bottom.h:64,
                 from ../src/rrd_config.h:634,
                 from ../src/compat-cloexec.h:4,
                 from test_compat-cloexec.c:1:
C:/msys64/mingw64/include/fcntl.h:36: note: this is the location of the previous definition
   36 | #define O_RDONLY _O_RDONLY
      | 
test_compat-cloexec.c:11: warning: "O_RDWR" redefined
   11 | #  define O_RDWR        0
      | 
C:/msys64/mingw64/include/fcntl.h:38: note: this is the location of the previous definition
   38 | #define O_RDWR _O_RDWR
      | 
test_compat-cloexec.c:12: warning: "O_APPEND" redefined
   12 | #  define O_APPEND      0
      | 
C:/msys64/mingw64/include/fcntl.h:39: note: this is the location of the previous definition
   39 | #define O_APPEND _O_APPEND
      | 
  CC       compat-cloexec.o
../src/compat-cloexec.c:11: warning: "O_CREAT" redefined
   11 | #  define O_CREAT       0
      | 
In file included from ../src/rrd_config_bottom.h:64,
                 from ../src/rrd_config.h:634,
                 from ../src/compat-cloexec.h:4,
                 from ../src/compat-cloexec.c:1:
C:/msys64/mingw64/include/fcntl.h:40: note: this is the location of the previous definition
   40 | #define O_CREAT _O_CREAT
      | 
../src/compat-cloexec.c:12: warning: "O_WRONLY" redefined
   12 | #  define O_WRONLY      0
      | 
C:/msys64/mingw64/include/fcntl.h:37: note: this is the location of the previous definition
   37 | #define O_WRONLY _O_WRONLY
      | 
../src/compat-cloexec.c:13: warning: "O_RDONLY" redefined
   13 | #  define O_RDONLY      0
      | 
C:/msys64/mingw64/include/fcntl.h:36: note: this is the location of the previous definition
   36 | #define O_RDONLY _O_RDONLY
      | 
../src/compat-cloexec.c:14: warning: "O_APPEND" redefined
   14 | #  define O_APPEND      0
      | 
C:/msys64/mingw64/include/fcntl.h:39: note: this is the location of the previous definition
   39 | #define O_APPEND _O_APPEND
      | 
../src/compat-cloexec.c:15: warning: "O_RDWR" redefined
   15 | #  define O_RDWR        0
      | 
C:/msys64/mingw64/include/fcntl.h:38: note: this is the location of the previous definition
   38 | #define O_RDWR _O_RDWR
      | 
  CCLD     compat-cloexec.exe
```

File: tests/test_compat-cloexec.c [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+#endif
+
+	fclose(f);
+}
+
+int main(void) {
+	FILE *f;
+
+	f = _rrd_fopen("/dev/null", "r");
+	check_file(f, O_RDONLY, false, __LINE__);
```
Comment (c72578) [inline tests/test_compat-cloexec.c:59]: The test fails under msys2 mingw-w64:
```sh
test_compat-cloexec.c:59 fopen() failed
FAIL compat-cloexec.exe (exit status: 3)
```

**Pull request merged**